synedrion/src/cggmp21/sigma/mod_.rs

```diff
@@ -135,34 +136,37 @@ impl<P: SchemeParams> ModProof<P> {
         }
     }
 
-    pub fn verify(
-        &self,
-        rng: &mut impl CryptoRngCore,
-        pk: &PublicKeyPaillier<P::Paillier>,
-        aux: &impl Hashable,
-    ) -> bool {
+    pub fn verify(&self, pk: &PublicKeyPaillier<P::Paillier>, aux: &impl Hashable) -> bool {
         let challenge = ModChallenge::new(pk, &self.commitment, aux);
         if challenge != self.challenge {
             return false;
         }
 
+        let mut reader = XofHasher::new_with_dst(b"P_mod RNG").chain(aux).finalize_to_reader();
+        let mut seed = <ChaCha8Rng as SeedableRng>::Seed::default();
+        reader.read(&mut seed);
+        let mut rng = ChaCha8Rng::from_seed(seed);
+
         // The paper requires checking that `N` is odd here,
         // but it is already an invariant of `PublicKeyPaillier`.
+        if (*pk.modulus()).is_prime_with_rng(&mut rng) {
+            return false;
+        }
 
-        // Note: I think we can get away with using the default RNG here
-        // since the result is RNG-independent (or at least supposed to be).
-        // It is possible to pass the external RNG similarly to how it's done for `new()`,
-        // but it would require quite a bit of changes because an external RNG is not accessible
-        // at the callsite.
-        if (*pk.modulus()).is_prime_with_rng(rng) {
+        if !self.commitment.0.is_invertible(pk.modulus()) {
             return false;
         }
 
         let monty_params = pk.monty_params_mod_n();
-        let omega_mod = self.commitment.0.to_montgomery(monty_params);
+        let w_mod = self.commitment.0.to_montgomery(monty_params);
         for (elem, y) in self.proof.iter().zip(self.challenge.0.iter()) {
             let z_m = elem.z.to_montgomery(monty_params);
             let mut y_m = y.to_montgomery(monty_params);
+
+            if !y.is_invertible(pk.modulus()) {
+                return false;
+            }
```

Comment:
Is this checking that yᵢ ∈ Z*_N, i.e. if it's not invertible it's not a part of the group?

Reply:
Yes, `Z*_N` is by definition the set of all invertible elements of Z_N (apparently so common of a notation they didn't even bother to mention it in the paper, but it is confirmed by Nikos and listed in `PAPER.md`)

Reply:
Worth a comment imo, just to make it easier to cross-ref with the paper.